_flush() - Wait for all promises to be settled

ro0gr opened the pull request:
Added a quick test for case when there are 2 `onExit` handlers defined.
The first `onExit` handler is a Promise **rejected** in 100ms.
The second one is a Promise **resolved**  in 1 sec.
I expect both of them to be fulfilled before process is finished.

If it's proven to be a real issue i would be glad to replace it with more precise unit test.

ro0gr committed:
_flush() unit test for promises with exit codes

rwjblue commented:
Yes, I think both should settle. I believe this is caused by https://github.com/ember-cli/capture-exit/blob/master/index.js#L109, swapping for `allSettled` seems like it would fix...

rwjblue commented:
Also, I think it should be possible to test this in the _flush unit tests (instead of the spawn style tests).

ro0gr commented:
@rwjblue thanks for your guidance.

Updated test per your suggestion. Will re-base once finish the work.

ro0gr committed:
Wait for all promises to be settled

ro0gr commented:
@rwjblue you right. `allSettled` makes the trick.
Rebased.

stefanpenner commented on index.js:
```diff
@@ -106,7 +106,7 @@ module.exports._flush = function(lastTime, code) {
   return RSVP.Promise.resolve(lastTime).
     then(function() {
-      return RSVP.Promise.all(work.map(function(handler) {
+      return RSVP.allSettled(work.map(function(handler) {
```
this isn't correct, as if any of them reject. We will now not correctly reject.

I am OK with exploring using `allSettled` so we wait for all the settle, but if any of them failed we should still reject the final promise. This gets tricky when it becomes a rollup of many rejections..

This is nuanced area, we should think through this carefully. But as currently written, legit failures will be hidden.

ro0gr replied:
Thanks for catching this. I'll think about it

ro0gr committed:
Re-throw the first rejected handler's exit code

ro0gr commented on test.js:
```diff
@@ -199,9 +234,34 @@ describe('capture-exit', function() {
+          throw new Error('Promise should be rejected');
+        })
+        .catch(function(reason) {
+          expect(reason).to.equal(400);
```
The third handler rejects faster then the second. Anyways registration order takes precedence here.

ro0gr commented:
I've updated implementation to [search for the first rejected promise and then re-throw](https://github.com/ember-cli/capture-exit/pull/18/files#diff-168726dbe96b3ce427e7fedce31bb0bcR112).

_One thing to note_: it re-throws not the first rejected handler but the first registered one which rejects. There is a [test](https://github.com/ember-cli/capture-exit/pull/18/files#diff-1dd241c4cd3fd1dd89c570cee98b79ddR258) for it.

@stefanpenner does this change address your concern about `allSettled` or am I still  missing something?

ro0gr commented on index.js:
```diff
@@ -106,9 +106,17 @@ module.exports._flush = function(lastTime, code) {
         return handler.call(null, code);
-      }));
+      })).then(function(results) {
+        var firstRejectedHandler = results.find(function(p) {
```
tests fail for `[].find()` with `node@0.12` do we still need to support it?

rwjblue replied:
We can probably remove support for it ([see here](https://github.com/ember-cli/ember-cli/pull/6622)), but it should probably be a "major version" bump (i.e. v2.0 here).

My suggestion would be to use `.filter(currentFunction)[0]` for now (so that we can get this fix released in v1.x) then replace (or open an issue so we don't forget) and release v2.x.

ro0gr commented:
@rwjblue replaced `find(fn)` with `filter(fn)[0]` per your suggestion and rebased. Will create an issue for `filter`=>`find` once this pr is accepted.

hjdivad commented on test.js:
```diff
@@ -199,9 +234,34 @@ describe('capture-exit', function() {
+      return exit._flush()
+        .then(function() {
+          throw new Error('Promise should be rejected');
+        })
```
This fulfillment handler is unnecessary: the handlers above are already rejecting the promise.

ro0gr replied:
Yes. I wanted to fail a test if the promised is resolved but instead introduced a bug here. Will fix, thanks

ro0gr replied:
fixed

Commit:
Fix test for failed handler exitCode

hjdivad commented:
Something I noticed while reviewing https://github.com/ember-cli/ember-cli/pull/6615 is that PR seems to expect rejecting handlers to be able to determine exit code.  I might be missing something, but it really looks like that [isn't the case](https://github.com/ember-cli/capture-exit/blob/5acaf39e3f84265dba9e8dbcc0da4727f582f85f/index.js#L91-L98).

If handlers reject with an integer, whether to use it for an exit code or not seems to have three cases:

1. `beforeExit`
2. `process.exit(0)`
3. `process.exit(nonzero)`

Having typed this out and thought about it in the process I think we don't want to support this because it's not obvious what the best thing to do is and the rules may get tricky.  That said, I'm interested to hear others' thoughts.

In the case where handlers want to control the exit code they already can, not by rejecting, but by calling (the captured) `process.exit(code)`.  This will generally behave as we want.

stefanpenner commented:
> Having typed this out and thought about it in the process I think we don't want to support this because it's not obvious what the best thing to do is and the rules may get tricky. That said, I'm interested to hear others' thoughts.

I suspect controlling the exit code is fraught with peril, but waiting for all pending tasks to complete before exiting seems ok no?

hjdivad commented:
> I suspect controlling the exit code is fraught with peril, but waiting for all pending tasks to complete before exiting seems ok no?

Yes I'm 👍  on this PR.

The other PR seems to expect to be able to `Promise.reject(code)` which is why I brought it up.

hjdivad commented:
The tests in this PR also kind of imply that handlers can set exit code but that seems like too minor of an issue to matter.

stefanpenner commented on test.js:
```diff
@@ -199,9 +234,38 @@ describe('capture-exit', function() {
       });
     });
 
+    it('throws an exit code of the first registered handler which rejects', function() {
```
I don't know about exit code preservation being tested. It seems to be that it important to test the following:

all handlers (even the very slow) are invoked before the real `process.exit` is called

ro0gr replied:
@stefanpenner If I understand you correctly, you mean a separate acceptance(`childProcess.execSync()`) test here. Right?

stefanpenner replied:
i believe this test can be implemented merely by spying on the real `process.exit` and ensuring it is called at the right and expected time.

ro0gr replied:
@stefanpenner added tests per your suggestion #21 
Please let me know if I misunderstood anything.

ro0gr commented:
Now after digging to tests I believe I understand better why exiting with a code provided by rejected promise is a problem.

Anyway from the [current implementation](https://github.com/ember-cli/capture-exit/blob/master/index.js#L94) I can see that currently `capture-exit` exits with an error provided by the first rejected(not registered) exit handler. Looks like it behaves in similar way as with calling of `process.exit(code)` manually.
But the original issue(waiting for promises are settled) remains. 

I can see 2 possible ways to move it forward:
 - get rid of promise rejection handling. Seems like it's undesirable change.
 - store rejection time of each exit handler and than throw the first rejected code

As for me the second option seems better. There is an opened question in case we choose this direction: should this behavior(preserving of rejection time and re-throwing) be replicated to `beforeExit`? 

@hjdivad @stefanpenner  your thoughts?

ro0gr committed:
update test to ensure that first rejeted promise is re-thrown

ro0gr committed:
re-throw the first rejected handler

ro0gr commented:
updated test to ensure that we re-throw not the first registered handler which rejects but the first rejected. Seems like this behavior is an original intention.

Also updated an implementation too match the original behavior.

stefanpenner commented on index.js:
```diff
@@ -106,9 +106,19 @@ module.exports._flush = function(lastTime, code) {
+        if (firstRejected) {
+          throw firstRejected;
+        }
+      });
```
previously this promise chain returned the fulfilled handler return values, it no longer does. 

I believe that can be solved by:

```js
return result.map(x => x.value); 
```

in this last `then` `fulfillment` handler.

ro0gr replied:
as I can see we have nothing to do with handler return values. Do we really need to care about it?

stefanpenner replied:
maybe not, just noticed it as a change.

ro0gr replied:
yeah. I had to provide a comment here  after change

ro0gr commented:
Thanks you all guys for help with this @rwjblue @hjdivad @stefanpenner

stefanpenner commented:
@ro0gr thanks for doing the heavy lifting!
